### PR TITLE
sound.cpp: Fix missed samples due to state save (#9917)

### PR DESCRIPTION
**Pull request by couriersud**
The PR address #9917. A save state may occur between time slices.
Sound devices through sound.cpp are updated during a timer call every
20ms. When the state is saved, these devices are not updated to the
current machine time. Consequently after a state load the devices
have a "time lag" since in postload buffer end time is forced to
machine time.
This change will save the last buffer end time so that all outstanding
samples are processed.

This is a core change. I tested it on some drivers. This needs a very
thorough review and I post the PR primarily to document a possible
solution.

### CODE REVIEW

File: src/emu/sound.cpp [hunk 1]
```diff
@@ -579,6 +580,7 @@ sound_stream::sound_stream(device_t &device, u32 inputs, u32 outputs, u32 output
 	std::string state_tag = string_format("%d", m_device.machine().sound().unique_id());
 	auto &save = m_device.machine().save();
 	save.save_item(&m_device, "stream.sample_rate", state_tag.c_str(), 0, NAME(m_sample_rate));
+	save.save_item(&m_device, "stream.sample_rate", state_tag.c_str(), 0, NAME(m_last_update_end));
```
Comment (galibert) [inline src/emu/sound.cpp:583]: Same name?  How the hell did that work?

File: src/emu/sound.cpp [hunk 2]
```diff
@@ -859,8 +866,13 @@ void sound_stream::postload()
...
 	for (auto &output : m_output)
-		output.set_end_time(m_device.machine().time());
-
+	{
+		// previous : machine time
+		// output.set_end_time(m_device.machine().time());
+		// printf("postload %s %f\n", name().c_str(), m_last_update_end.as_double());
+		// resampler is zero
+		output.set_end_time(m_last_update_end > attotime::zero ? m_last_update_end : m_device.machine().time());
+	}
```
Comment (galibert) [inline src/emu/sound.cpp:875]: I don't understand.  output has a end time stored somewhere in some form, why don't you save that?

File: src/emu/sound.cpp (hunk 1 above)
Comment (couriersud) [reply src/emu/sound.cpp:583]: I assume it's `NAME(m_last_update_end)` as the last parameter. I'll fix that.

File: src/emu/sound.cpp (hunk 2 above)
Comment (couriersud) [reply src/emu/sound.cpp:875]: I would need a fixed array to cache the value of `output.end_time()`. And in addition a `pre_save` handler. The end time for all outputs is always the same. That's why I used this approach.
In addition there is stuff like (line 680)
```
	// ignore any update requests if we're already up to date
	attotime start = m_output[0].end_time();
	attotime end = m_device.machine().time();
	if (start >= end)
	{
		m_last_update_end = start;
		return;
	}

	// regular update then
	update_view(start, end);
```
This indicates that the correct sound device time is end_time() for `start >= end`. In the other case it is `machine().time()`.

File: src/emu/sound.cpp (hunk 1 above)
Comment (couriersud) [reply src/emu/sound.cpp:583]: Here is another one:
```
void sound_stream_input::init(sound_stream &stream, u32 index, char const *tag, sound_stream_output *resampler)
{
	// set the passed-in values
	m_owner = &stream;
	m_index = index;
	m_resampler_source = resampler;

	// save our state
	auto &save = stream.device().machine().save();
	save.save_item(&stream.device(), "stream.input", tag, index, NAME(m_gain));
	save.save_item(&stream.device(), "stream.input", tag, index, NAME(m_user_gain));
}
```
There it more looks like that the code should be something like
```
	save.save_item(&m_device, "stream.base_stream", state_tag.c_str(), 0, NAME(m_sample_rate));
	save.save_item(&m_device, "stream.base_stream", state_tag.c_str(), 0, NAME(m_last_update_end));
```

**Pull request merged**